Fix KV cache sync issue during CUDA graph replay

yashwantbezawada opened the pull request:
Fixes #29608

I ran into this while looking at the async KV transfer code. When using connectors like LMCache with CUDA graphs, there's a race condition during graph replay.

The issue is that `wait_for_layer_load()` calls happen inside the `@maybe_transfer_kv_layer` decorator on attention functions. During graph capture, these calls execute normally. But during replay, only the GPU operations are replayed - the Python decorator code gets skipped entirely. So the async KV loads might not be complete when the attention kernels run.

The fix adds a `wait_for_load()` method to the connector interface that waits for all pending loads to complete. I put the sync call right before `replay()` in CUDAGraphWrapper so it only runs during replay, not capture. During capture we still want the per-layer sync to work normally so saves happen correctly.

Changes:
- Added `wait_for_load()` to KVConnectorBase_V1 with a default no-op (connectors that don't do async loading are unaffected)
- Implemented it in LMCacheConnectorV1Impl to drain any remaining layers from the retrievers
- Added it to MultiConnector to delegate to sub-connectors
- Added the sync call in CUDAGraphWrapper before replay()

I couldn't test this end-to-end without an LMCache setup, but traced through the code paths carefully. The sync only happens when both KV transfer and CUDA graphs are enabled, and only during replay.

yashwantbezawada committed:
Fix KV cache sync issue during CUDA graph replay

When using async KV transfer connectors with CUDA graphs, the per-layer
wait_for_layer_load() calls in the attention decorator are bypassed
during graph replay, causing a race condition where attention kernels
can execute with incomplete KV cache data.

Added wait_for_load() method to KVConnectorBase_V1 that ensures all
async KV loads complete before CUDA graph replay. The sync is placed
right before replay() in CUDAGraphWrapper so it only affects replay,
not capture.

Fixes #29608

Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

ProExpertProg reviewed (changes requested):
also cc @KuntaiDu @NickLucche

NickLucche reviewed (commented):
I think this may actually be related to the changes introduced in https://github.com/vllm-project/vllm/pull/28253 which intentionally skips the tracing.
There's a second PR to address that here https://github.com/vllm-project/vllm/pull/27026 by @sdavidbd .

Does that match your observations @yashwantbezawada ?

yashwantbezawada commented:
Thanks for the context!

Looked through #28253 and #27026 - those are fixing the capture/dummy run path (getting metadata set properly so KV connector works during capture).

This one's targeting a different scenario: during replay, `start_load_kv()` runs (it's in the model runner, outside the graph), but the decorator's `wait_for_layer_load()` doesn't execute since only captured GPU ops run. So no sync between the async loads and replayed attention.

Fix adds sync in the model runner after `start_load_kv()`, before the forward.

Think they should be complementary - #27026 for capture, this for replay. But happy to chat more or coordinate with @sdavidbd if there's overlap!

ProExpertProg reviewed (approved):
Approving to remove the "requested changes" mark while I'm gone for the next 2 weeks, please wait for @NickLucche's approval to merge!

yashwantbezawada committed:
Update vllm/v1/worker/kv_connector_model_runner_mixin.py

Co-authored-by: Luka Govedič <ProExpertProg@users.noreply.github.com>
Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

NickLucche reviewed (commented):
Hey @yashwantbezawada sorry for the delay in getting back to this. 
I think there's some confusion around `start_load_kv` as you also noted in your comment, perhaps @ApostaC can chime in to clarify the original intent.
`start_load_kv` is supposed to start an (*async*) transfer of all layers. 
So far there hasn't been the need to explicitly sync on this transfer, otherwise we would lose the benefit of real async (wrt engine execution) transfer.
Now, if a sync was really needed for any reason at all, one could overload `wait_for_save`. If this was the case, I would agree resorting to your method would offer better semantics, but this hasn't been the case afaik.

There's also a second set of operations that can be used to move kv cache layer-by-laeyr and that naturally go together <save_kv_layer, wait_for_layer_load>.
If tracing of these ops is skipped, then they will be indeed skipped from replay. But I think this is a separate issue that we should address by either a) tracing the save/load b) only allow their usage in piecewise.

tl;dr: I think we're fixing a separate part of the interface to make up for a potential logic bug in how we're recording ops for kv cache save/load.

yashwantbezawada committed:
Update vllm/v1/worker/kv_connector_model_runner_mixin.py

Co-authored-by: Nicolò Lucchesi <nicolo.lucchesi@gmail.com>
Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

yashwantbezawada commented on vllm/v1/worker/kv_connector_model_runner_mixin.py:
```diff
@@ -126,7 +126,15 @@ def _get_kv_connector_output(
+        # capture/replay. During replay, the per-layer wait_for_layer_load()
+        # calls in the attention layer decorator are bypassed.
+        if forward_ctx.cudagraph_runtime_mode == CUDAGraphMode.FULL:
+            kv_connector.wait_for_load()
```
Thanks for the detailed feedback.

You're right - `start_load_kv()` is designed for layer-by-layer async loading in parallel with attention, and adding `wait_for_load()` as a global sync would eliminate that concurrency benefit.

Your suggestion makes sense: if a connector needs full sync for CUDA graphs, it could handle that internally in `start_load_kv()` by checking `forward_ctx`. That keeps the decision with each connector rather than imposing uniform semantics via a new API.

Happy to close this PR or rework it based on your guidance. What would you recommend?

yashwantbezawada replied:
Done! Moved the sync logic into LMCache's `start_load_kv()` - it now drains all layer retrievers at the end when CUDA graphs are enabled. Removed the `wait_for_load()` API entirely.

yashwantbezawada committed:
Move KV sync into connector's start_load_kv() for CUDA graphs

Instead of adding a new wait_for_load() API, move the sync decision
into the connector itself. LMCache now drains all layer retrievers
at the end of start_load_kv() when CUDA graphs are enabled.

This keeps the decision with each connector rather than imposing
uniform sync semantics via a global API.

Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>
